### PR TITLE
add tests for TA.prototype.with/filter shrinking/detaching buffer during coercion

### PR DESCRIPTION
**Pull request by bakkot**
For https://github.com/tc39/ecma262/pull/3979. Written by an LLM based off that PR.

These pass in the engines I have on hand except that some of them fail in Boa, which asserts as the spec currently (and incorrectly) does.

Also added the test/built-ins/TypedArray/prototype/with/BigInt/negative-index-resize-to-out-of-bounds.js because my version of Node throws the wrong error in that case for some reason (but v8 gets it right).

### CODE REVIEW

**Comment by bakkot** [comment]
Huh, surprised by the jsc failures; it works on my machine.

File: test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer-kept-undefined.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+  assert.throws(TypeError, function() {
+    sample.filter(function(value, index, array) {
+      assert.sameValue(index, calls);
+      assert.sameValue(array, sample);
+      if (calls === 0) {
+        assert.sameValue(value, 1n);
+        $DETACHBUFFER(sample.buffer);
+      } else {
+        assert.sameValue(value, undefined, "Missing elements are passed to the callback");
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer-kept-undefined.js]:
```suggestion
        assert.sameValue(value, undefined,
          "callback element is undefined for index " + index);
```

File: test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer.js [hunk 2]
```diff
@@ -14,24 +14,27 @@ info: |
...
 
-  sample.filter(function() {
+  var result = sample.filter(function(value) {
     var flag = true;
     if (loops === 0) {
+      assert.sameValue(value, 0n);
       $DETACHBUFFER(sample.buffer);
     } else {
-      flag = false;
+      assert.sameValue(value, undefined, "Missing elements are passed to the callback");
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer.js]:
```suggestion
      assert.sameValue(value, undefined,
        "callback element is undefined for index " + index);
```

File: test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-shrink-kept-undefined.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    assert.throws(TypeError, function() {
+      sample.filter(function(value, index, array) {
+        assert.sameValue(index, calls);
+        assert.sameValue(array, sample);
+        if (calls === 0) {
+          assert.sameValue(value, 1n);
+          buffer.resize(TA.BYTES_PER_ELEMENT);
+        } else {
+          assert.sameValue(value, undefined, "Missing elements are passed to the callback");
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-shrink-kept-undefined.js]:
```suggestion
          assert.sameValue(value, undefined,
            "callback element is undefined for index " + index);
```

File: test/built-ins/TypedArray/prototype/filter/callbackfn-detachbuffer.js [hunk 4]
```diff
@@ -13,21 +13,26 @@ info: |
...
   var loops = 0;
   var sample = new TA(makeCtorArg(2));
 
-  sample.filter(function() {
+  var result = sample.filter(function(value) {
     if (loops === 0) {
+      assert.sameValue(value, 0);
       $DETACHBUFFER(sample.buffer);
+    } else {
+      assert.sameValue(value, undefined, "Missing elements are passed to the callback");
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/filter/callbackfn-detachbuffer.js]:
```suggestion
      assert.sameValue(value, undefined,
        "callback element is undefined for index " + index);
```

File: test/built-ins/TypedArray/prototype/with/BigInt/index-coercion-shrinks.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  };
+
+  // Index 0 remains valid, but copying the remaining elements reads undefined.
+  assert.throws(TypeError, function() {
+    sample.with(index, 9n);
+  });
+  assert.sameValue(calls, 1, "index is coerced once");
+  assert.sameValue(sample.length, 1);
+  assert.sameValue(sample[0], 1n);
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/with/BigInt/index-coercion-shrinks.js]:
```suggestion
  assert.compareArray(sample, [1n]);
```

File: test/built-ins/TypedArray/prototype/with/BigInt/value-coercion-shrinks.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  };
+
+  // Index 0 remains valid, but copying the remaining elements reads undefined.
+  assert.throws(TypeError, function() {
+    sample.with(0, value);
+  });
+  assert.sameValue(calls, 1, "value is coerced once");
+  assert.sameValue(sample.length, 1);
+  assert.sameValue(sample[0], 1n);
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/with/BigInt/value-coercion-shrinks.js]:
```suggestion
  assert.compareArray(sample, [1n]);
```

File: test/built-ins/TypedArray/prototype/with/index-coercion-shrinks.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  };
+
+  // Index 0 remains valid, but copying the remaining elements reads undefined.
+  var result = sample.with(index, 9);
+  var converted = isFloatTypedArrayConstructor(TA) ? NaN : 0;
+  assert.compareArray(result, [9, converted, converted]);
+  assert.sameValue(calls, 1, "index is coerced once");
+  assert.sameValue(sample.length, 1);
+  assert.sameValue(sample[0], 1);
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/with/index-coercion-shrinks.js]:
```suggestion
  assert.compareArray(sample, [1]);
```

File: test/built-ins/TypedArray/prototype/with/value-coercion-shrinks.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  };
+
+  // Index 0 remains valid, but copying the remaining elements reads undefined.
+  var result = sample.with(0, value);
+  var converted = isFloatTypedArrayConstructor(TA) ? NaN : 0;
+  assert.compareArray(result, [9, converted, converted]);
+  assert.sameValue(calls, 1, "value is coerced once");
+  assert.sameValue(sample.length, 1);
+  assert.sameValue(sample[0], 1);
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/with/value-coercion-shrinks.js]:
```suggestion
  assert.compareArray(sample, [1]);
```

**Review by gibson042 — APPROVED** [review]

File: test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer.js [hunk 9]
```diff
@@ -14,24 +14,28 @@ info: |
...
-  sample.filter(function() {
+  var result = sample.filter(function(value) {
     var flag = true;
     if (loops === 0) {
+      assert.sameValue(value, 0n);
       $DETACHBUFFER(sample.buffer);
     } else {
-      flag = false;
+      assert.sameValue(value, undefined,
+        "callback element is undefined for index " + index);
```
Comment (linusg) [inline test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer.js:32]: `index` is not defined

File: test/built-ins/TypedArray/prototype/filter/callbackfn-detachbuffer.js [hunk 10]
```diff
@@ -13,21 +13,27 @@ info: |
...
   var sample = new TA(makeCtorArg(2));
 
-  sample.filter(function() {
+  var result = sample.filter(function(value) {
     if (loops === 0) {
+      assert.sameValue(value, 0);
       $DETACHBUFFER(sample.buffer);
+    } else {
+      assert.sameValue(value, undefined,
+        "callback element is undefined for index " + index);
```
Comment (linusg) [inline test/built-ins/TypedArray/prototype/filter/callbackfn-detachbuffer.js:30]: `index` is not defined

**Review by linusg — CHANGES_REQUESTED** [review]

File: test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer.js [hunk 11]
```diff
@@ -14,24 +14,28 @@ info: |
...
+includes: [detachArrayBuffer.js, testTypedArray.js, compareArray.js]
 features: [BigInt, TypedArray]
 ---*/
 
 testWithBigIntTypedArrayConstructors(function(TA, makeCtorArg) {
   var loops = 0;
   var sample = new TA(makeCtorArg(2));
 
-  sample.filter(function() {
+  var result = sample.filter(function(value) {
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/filter/BigInt/callbackfn-detachbuffer.js]:
```suggestion
  var result = sample.filter(function(value, index) {
```

File: test/built-ins/TypedArray/prototype/filter/callbackfn-detachbuffer.js [hunk 12]
```diff
@@ -13,21 +13,27 @@ info: |
...
+includes: [detachArrayBuffer.js, testTypedArray.js, compareArray.js]
 features: [TypedArray]
 ---*/
 
 testWithTypedArrayConstructors(function(TA, makeCtorArg) {
   var loops = 0;
   var sample = new TA(makeCtorArg(2));
 
-  sample.filter(function() {
+  var result = sample.filter(function(value) {
```
Comment (gibson042) [inline test/built-ins/TypedArray/prototype/filter/callbackfn-detachbuffer.js]:
```suggestion
  var result = sample.filter(function(value, index) {
```

**Review by linusg — APPROVED** [review]
